pkg/kv/kvnemesis/kvnemesis_test.go

```diff
@@ -35,33 +42,99 @@ func init() {
 	numSteps = envutil.EnvOrDefaultInt("COCKROACH_KVNEMESIS_STEPS", 50)
 }
 
-func TestKVNemesisSingleNode(t *testing.T) {
-	defer leaktest.AfterTest(t)()
-	defer log.Scope(t).Close(t)
-	skip.UnderRace(t)
+func testClusterArgs(tr *SeqTracker) base.TestClusterArgs {
+	storeKnobs := &kvserver.StoreTestingKnobs{
+		// Drop the clock MaxOffset to reduce commit-wait time for
+		// transactions that write to global_read ranges.
+		MaxOffset: 10 * time.Millisecond,
+		// Make sure we know the seq for each of our writes when they come out of
+		// the rangefeed. We do this via an interceptor to avoid having to change
+		// RangeFeed's APIs.
+		RangefeedValueHeaderFilter: func(key, endKey roachpb.Key, ts hlc.Timestamp, vh enginepb.MVCCValueHeader) {
+			if seq := kvnemesisutil.Seq(vh.KVNemesisSeq.Get()); seq > 0 {
+				tr.Add(key, endKey, ts, seq)
+			}
+		},
+	}
 
-	ctx := context.Background()
-	tc := testcluster.StartTestCluster(t, 1, base.TestClusterArgs{
+	return base.TestClusterArgs{
 		ServerArgs: base.TestServerArgs{
 			Knobs: base.TestingKnobs{
-				Store: &kvserver.StoreTestingKnobs{
-					// Drop the clock MaxOffset to reduce commit-wait time for
-					// transactions that write to global_read ranges.
-					MaxOffset: 10 * time.Millisecond,
+				Store: storeKnobs,
+				KVClient: &kvcoord.ClientTestingKnobs{
+					// Don't let DistSender split DeleteRangeUsingTombstone across range boundaries.
+					// This does happen in real CRDB, but leads to separate atomic subunits, which
+					// would add complexity to kvnemesis that isn't worth it. Instead, the operation
+					// generator for the most part tries to avoid range-spanning requests, and the
+					// ones that do end up happening get a hard error.
+					OnRangeSpanningNonTxnalBatch: func(ba *roachpb.BatchRequest) *roachpb.Error {
+						for _, req := range ba.Requests {
+							if req.GetInner().Method() != roachpb.DeleteRange {
+								continue
+							}
+							if req.GetDeleteRange().UseRangeTombstone == true {
+								return roachpb.NewErrorf("DeleteRangeUsingTombstone can not straddle range boundary")
+							}
+						}
+						return nil
+					},
 				},
 			},
 		},
-	})
+	}
+}
+
+func randWithSeed(
+	t interface {
+		Logf(string, ...interface{})
+		Helper()
+	}, seedOrZero int64,
+) *rand.Rand {
+	t.Helper()
+	var rng *rand.Rand
+	if seedOrZero > 0 {
+		rng = rand.New(rand.NewSource(seedOrZero))
+	} else {
+		rng, seedOrZero = randutil.NewTestRand()
+	}
+	t.Logf("seed: %d", seedOrZero)
+	return rng
+}
+
+func TestKVNemesisSingleNode(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+	skip.UnderRace(t)
+
+	if !buildutil.CrdbTestBuild {
+		// `roachpb.RequestHeader` and `MVCCValueHeader` have a KVNemesisSeq field
+		// that is zero-sized outside of test builds. We could revisit that should
+		// a need arise to run kvnemesis against production binaries.
+		skip.IgnoreLint(t, "kvnemesis must be run with the crdb_test build tag")
```

Comment:
nit: this is slightly annoying, since I can't trivially run kvnemesis under VSCode (nor in a debugger) without fiddling with build tag settings. But given that Go doesn't expose the equivalent of a test build tag, I don't think we can do much else? I think trading off test convenience for production performance is the right call here though.